crates/driver/src/domain/competition/solution/scoring.rs

```
@@ -158,17 +158,84 @@ impl Trade {
+            if !i.is_zero() {
+                current_trade.custom_price = Self::calculate_custom_prices(
+                    self.fulfillment.side(),
+                    self.fulfillment.executed().into(),
+                    self.fulfillment.fee().0.into(),
+                    &self.uniform_prices,
+                    amount,
+                )
+                .map_err(|e| Error::CustomPrice(e.to_string()))?;
+            }
         }
 
-        let protocol_fee = |policy: &order::FeePolicy| match policy {
-            order::FeePolicy::Surplus {
+        Ok(eth::Asset {
+            token: self.surplus_token(),
+            amount,
+        })
+    }
+
+    pub fn calculate_custom_prices(
+        side: Side,
+        executed: TokenAmount,
+        fulfillment_fee: TokenAmount,
+        uniform_prices: &ClearingPrices,
+        current_protocol_fee: TokenAmount,
+    ) -> Result<CustomClearingPrices, error::Scoring> {
+        Ok(CustomClearingPrices {
+            sell: match side {
+                Side::Sell => executed
+                    .0
+                    .checked_mul(uniform_prices.sell)
+                    .ok_or(Math::Overflow)?
+                    .checked_ceil_div(&uniform_prices.buy)
+                    .ok_or(Math::DivisionByZero)?
+                    .checked_add(current_protocol_fee.0)
+                    .ok_or(Math::Overflow)?,
+                Side::Buy => executed.0,
+            },
+            buy: match side {
+                Side::Sell => executed.0 + fulfillment_fee.0,
+                Side::Buy => (executed.0)
+                    .checked_mul(uniform_prices.buy)
+                    .ok_or(Math::Overflow)?
+                    .checked_div(uniform_prices.sell)
+                    .ok_or(Math::DivisionByZero)?
+                    .checked_add(fulfillment_fee.0)
+                    .ok_or(Math::Overflow)?
+                    .checked_sub(current_protocol_fee.0)
```

Why do we need to return an error here instead of using `saturating_sub`?